File: 2026POTD/JANUARY/13-01-2026/3453.go
func separateSquares(squares [][]int) float64 {
    lo, hi := 1e18, -1e18
    total := 0.0
    for _, s := range squares {
        y := float64(s[1])
        l := float64(s[2])
        if y < lo {
            lo = y
        }
        if y+l > hi {
            hi = y + l
        }
        total += l * l
    }
    target := total / 2.0
    for i := 0; i < 80; i++ {
        mid := (lo + hi) * 0.5
        below := 0.0
        for _, s := range squares {
            y := float64(s[1])
            l := float64(s[2])
            if mid > y {
                h := mid - y
                if h > l {
                    h = l
                }
                below += h * l
            }
        }
        if below < target {
            lo = mid
        } else {
            hi = mid
        }
    }
    return lo
}

File: 2026POTD/JANUARY/13-01-2026/3453.php
class Solution {

    /**
     * @param Integer[][] $squares
     * @return Float
     */
    function separateSquares($squares) {
        $events = [];

        foreach ($squares as $s) {
            $y = (float)$s[1];
            $l = (float)$s[2];
            $events[] = [$y, $l];
            $events[] = [$y + $l, -$l];
        }

        usort($events, function($a, $b) {
            return $a[0] <=> $b[0];
        });

        $segments = [];
        $curWidth = 0.0;
        $prevY = $events[0][0];
        $totalArea = 0.0;

        foreach ($events as [$y, $delta]) {
            if ($y > $prevY && $curWidth > 0) {
                $area = ($y - $prevY) * $curWidth;
                $segments[] = [$prevY, $y, $curWidth, $area];
                $totalArea += $area;
            }
            $curWidth += $delta;
            $prevY = $y;
        }

        $target = $totalArea / 2.0;
        $acc = 0.0;

        foreach ($segments as [$y1, $y2, $w, $area]) {
            if ($acc + $area >= $target) {
                return $y1 + ($target - $acc) / $w;
            }
            $acc += $area;
        }

        return 0.0;
    }
}